File: db4o.net/Db4objects.Db4o.Tests/Db4objects.Db4o.Tests/Common/CS/StandaloneCSTestCaseBase.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

#if !SILVERLIGHT
using System.IO;
using Db4oUnit;
using Db4oUnit.Extensions;
using Db4objects.Db4o;
using Db4objects.Db4o.CS;
using Db4objects.Db4o.CS.Config;
using Db4objects.Db4o.CS.Internal;
using Db4objects.Db4o.CS.Internal.Config;
using Db4objects.Db4o.Config;
using Db4objects.Db4o.Foundation.IO;

namespace Db4objects.Db4o.Tests.Common.CS
{
	public abstract class StandaloneCSTestCaseBase : ITestCase
	{
		private int _port;

		public sealed class Item
		{
			// TODO fix db4ounit call logic - this should actually be run in C/S mode
		}

		/// <exception cref="System.Exception"></exception>
		public virtual void Test()
		{
			string fileName = DatabaseFile();
			File4.Delete(fileName);
			IObjectServer server = Db4oClientServer.OpenServer(CreateServerConfiguration(), fileName
				, -1);
			_port = server.Ext().Port();
			try
			{
				server.GrantAccess("db4o", "db4o");
				RunTest();
			}
			finally
			{
				server.Close();
				File4.Delete(fileName);
			}
		}

		private IServerConfiguration CreateServerConfiguration()
		{
			return Db4oClientServerLegacyConfigurationBridge.AsServerConfiguration(CreateConfiguration
				());
		}

		private IConfiguration CreateConfiguration()
		{
			IConfiguration config = Db4oFactory.NewConfiguration();
			Configure(config);
			return config;
		}

		/// <exception cref="System.Exception"></exception>
		protected virtual void WithClient(IContainerBlock block)
		{
			ContainerServices.WithContainer(OpenClient(), block);
		}

		protected virtual ClientObjectContainer OpenClient()
		{
			return (ClientObjectContainer)Db4oClientServer.OpenClient(CreateClientConfiguration
				(), "localhost", _port, "db4o", "db4o");
		}

		private IClientConfiguration CreateClientConfiguration()
		{
			return Db4oClientServerLegacyConfigurationBridge.AsClientConfiguration(CreateConfiguration
				());
		}

		protected virtual int Port()
		{
			return _port;
		}

		/// <exception cref="System.Exception"></exception>
		protected abstract void RunTest();

		protected abstract void Configure(IConfiguration config);

		private string DatabaseFile()
		{
			return Path.Combine(Path.GetTempPath(), "cc.db4o");
		}
	}
}
#endif // !SILVERLIGHT

File: db4oj.tests/src/com/db4o/db4ounit/common/cs/StandaloneCSTestCaseBase.java
/* Copyright (C) 2007  Versant Inc.  http://www.db4o.com */

package com.db4o.db4ounit.common.cs;

import com.db4o.*;
import com.db4o.config.*;
import com.db4o.cs.*;
import com.db4o.cs.config.*;
import com.db4o.cs.internal.*;
import com.db4o.cs.internal.config.*;
import com.db4o.foundation.io.*;

import db4ounit.*;
import db4ounit.extensions.*;

// TODO fix db4ounit call logic - this should actually be run in C/S mode
public abstract class StandaloneCSTestCaseBase implements TestCase {

	private int _port;

	public static final class Item {
	}
	
	public void test() throws Throwable {
		String fileName = databaseFile();
		File4.delete(fileName);
		final ObjectServer server = Db4oClientServer.openServer(createServerConfiguration(), fileName, -1);
		_port = server.ext().port();
		try {
			server.grantAccess("db4o", "db4o");
			
			runTest();
			
		} finally {
			server.close();
			File4.delete(fileName);
		}
	}

	private ServerConfiguration createServerConfiguration() {
		return Db4oClientServerLegacyConfigurationBridge.asServerConfiguration(createConfiguration());
	}

	private Configuration createConfiguration() {
		final Configuration config = Db4o.newConfiguration();
		configure(config);
		return config;
	}

	protected void withClient(ContainerBlock block) throws Throwable {
		ContainerServices.withContainer(openClient(), block);
	}

	protected ClientObjectContainer openClient() {
		return (ClientObjectContainer)Db4oClientServer.openClient(createClientConfiguration(), "localhost", _port, "db4o", "db4o");
	}

	private ClientConfiguration createClientConfiguration() {
		return Db4oClientServerLegacyConfigurationBridge.asClientConfiguration(createConfiguration());
	}

	protected int port() {
		return _port;
	}
	
	protected abstract void runTest() throws Throwable;

	protected abstract void configure(Configuration config);

	private String databaseFile() {
		return Path4.combine(Path4.getTempPath(), "cc.db4o");
	}

}